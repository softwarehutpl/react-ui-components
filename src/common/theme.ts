const theme: { [key: string]: { [key: string]: { [key: string]: string } } } = {
  colors: {
    primary: {
      base: '#000000',
      light: '#ffffff',
      dark: '#a8b3bd',
    },
    secondary: {
      base: '#3690e3',
      light: '#75b0e6',
      dark: '#034d91',
    },
    error: {
      base: '#f44336',
      light: '#f56c62',
      dark: '#a1140a',
    },
    warning: {
      base: '#faa72d',
      light: '#ffc36b',
      dark: '#995b00',
    },
    info: {
      base: '#7979d4',
      light: '#9797d1',
      dark: '#4c4cc7',
    },
    success: {
      base: '#4caf50',
      light: '#88c28a',
      dark: '#124f14',
    },
    disabled: {
      base: '#d9d9d9',
      light: '#f0eded',
      dark: '#8c8989',
    },
  },
  // TODO: change it in dedicated task for that (change ease-out for example and add more options)
  transitions: {
    all: {
      fast: 'all 0.5s ease-out',
      mid: 'all 2s ease-out',
      slow: 'all 4s ease-out',
    },
    background: {
      fast: 'background-color 0.5s ease-out',
      mid: 'background-color 2s ease-out',
      slow: 'background-color 4s ease-out',
    },
    fontColor: {
      fast: 'color 0.5s ease-out',
      mid: 'color 2s ease-out',
      slow: 'color 4s ease-out',
    }
  }
};

export default theme;
