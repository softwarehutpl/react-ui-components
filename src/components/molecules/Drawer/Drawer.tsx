import React from 'react';
import styled  from 'styled-components';
import { TRANSFORMS } from '../../../common/consts/consts';

export interface DrawerOverlayProps {
  size?: string;
  isOpen: boolean;
};

export interface DrawerContentProps {
  placement?: string;
  size?: string;
  color?: string;
  isOpen: boolean;
  transitionDuration?: string;
};

export interface DrawerProps extends DrawerOverlayProps, DrawerContentProps {
  className?: string;
  children?: any;
  onClose?: () => void;
};

const sizes = {
  large: '50%',
  medium: '40%',
  small: '30%',
} as any;

const placements = {
  top: {
    top: 0,
    right: 0,
    left: 0
  },
  right: {
    top: 0,
    right: 0,
    bottom: 0
  },
  bottom: {
    right: 0,
    bottom: 0,
    left: 0
  },
  left: {
    top: 0,
    bottom: 0,
    left: 0
  }
} as any;

export const DrawerOverlay = styled.div<DrawerOverlayProps>`
  position: fixed;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  background-color: ${({  theme }) =>  (theme.colors.disabled.base)};
  opacity: 0.3;
  transition: opacity .3s ease-in;
  display: ${({ isOpen }) => (isOpen ? 'block' : 'none')};
`;

export const DrawerContent = styled.div<DrawerContentProps>`
  background-color: ${({ theme, color }) => (color && theme.colors[color].light)};
  color: ${({ theme, color }) => (color && theme.colors[color].base)};
  position: fixed;
  ${({ placement }) => (placement && placements[placement])};
  width: ${({ placement, size }) => ((placement !== 'top' && placement !== 'bottom') && size) ? sizes[size] : size};
  height: ${({ placement, size }) => ((placement === 'top' || placement === 'bottom') && size) ? sizes[size] : size};
  transform: ${({ isOpen, placement }) => ((!isOpen && placement) ? TRANSFORMS[placement] : null)};
  transition:${({ transitionDuration }) => (`transform ${transitionDuration} ease-out`)};
  overflow-x: hidden;
  overflow-y: scroll;
  box-shadow: 0px 0px 10px rgba(0,0,0,0.12);
`;

const Drawer: React.StatelessComponent<DrawerProps> = ({
  className,
  children,
  placement,
  size,
  color,
  isOpen,
  onClose,
  transitionDuration
}) => (
  <>
    <DrawerOverlay 
      isOpen={isOpen}
      onClick={onClose}
    />
    <DrawerContent
      className={className}
      placement={placement}
      size={size}
      color={color}
      isOpen={isOpen}
      transitionDuration={transitionDuration}
    >
      {children}
    </DrawerContent>
  </>
);

Drawer.defaultProps = {
  color: 'primary',
  isOpen: false,
  placement: 'right',
  size: 'medium',
  transitionDuration: '0.2s',
};

export default Drawer;
