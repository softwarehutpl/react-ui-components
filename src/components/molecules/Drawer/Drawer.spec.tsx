import React from 'react';
import { create, act } from 'react-test-renderer';
import 'jest-styled-components';
import Drawer, { DrawerContent, DrawerOverlay } from './Drawer';
import { ThemeProvider } from 'styled-components';
import theme from '../../../common/theme';

describe('<Drawer />', () => {
  it('should render corectly', () => {
    const component = create(
      <ThemeProvider 
        theme={theme}
      >
        <Drawer />
      </ThemeProvider>
    );
    expect(component).toMatchSnapshot();
  });

  it('should render with className if provied', () => {
    const component = create(<DrawerContent className="test-class" />);
    const instance = component.root;
    instance.findByType('div');
    expect(instance.props.className).toEqual(expect.stringContaining('test-class'));
  });

  it('should render with children', () => {
    const component = create(<DrawerContent>child</DrawerContent>);
    const instance = component.root;
    instance.findByType('div');
    expect(instance.props.children).toBe('child');
  });

  it('should call button onClick method', () => {
    const mockOnClick = jest.fn();
    const component = create(<DrawerOverlay onClick={mockOnClick} theme={theme} />);
    const instance = component.root;
    const overlay = instance.findByType('div');
    act(() => {
      overlay.props.onClick();
    });
    expect(mockOnClick.mock.calls.length).toEqual(1);
  });

  it('should render with propery color', () => {
    const component = create(<DrawerContent color="primary" theme={theme} />).toJSON();
    expect(component).toHaveStyleRule('background-color', '#ffffff');
    expect(component).toHaveStyleRule('color', '#000000');
  });

  it('should have proper style with right placement prop', () => {
    const root  = create(
      <DrawerContent
        color="primary" 
        theme={theme} 
        size="large"
        placement="right"
        isOpen={false}
      />
    );

    const component = root.toJSON();
    expect(component).toHaveStyleRule('right', '0');
    expect(component).toHaveStyleRule('transform', 'translateX(100%)');   
    
  });

  it('should have proper style with left placement prop', () => {
    const root  = create(
      <DrawerContent
        color="primary" 
        theme={theme} 
        size="large"
        placement="left"
        isOpen={false}
      />
    );

    const component = root.toJSON();
    expect(component).toHaveStyleRule('left', '0');
    expect(component).toHaveStyleRule('transform', 'translateX(-100%)'); 
    
  });

  it('should have proper style with top placement prop', () => {
    const root  = create(
      <DrawerContent
        color="primary" 
        theme={theme} 
        size="large"
        placement="top"
        isOpen={false}
      />
    );

    const component = root.toJSON();
    expect(component).toHaveStyleRule('top', '0');
    expect(component).toHaveStyleRule('transform', 'translateY(-100%)')
    
  });

  it('should have proper style with bottom placement prop', () => {
    const root  = create(
      <DrawerContent
        color="primary" 
        theme={theme} 
        size="large"
        placement="bottom"
        isOpen={false}
      />
    );

    const component = root.toJSON();
    expect(component).toHaveStyleRule('bottom', '0');
    expect(component).toHaveStyleRule('transform', 'translateY(100%)');   
  });

  it('should have proper style when isOpen prop is provided', () => {
    const component  = create(
      <DrawerOverlay
        theme={theme}
        color="primary" 
        isOpen
      />
    ).toJSON();

    expect(component).toHaveStyleRule('display', 'block');
  });
});
