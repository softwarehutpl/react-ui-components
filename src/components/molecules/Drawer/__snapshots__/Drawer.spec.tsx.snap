// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Drawer /> should render corectly 1`] = `
Array [
  .c0 {
  position: fixed;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  background-color: #d9d9d9;
  opacity: 0.3;
  -webkit-transition: opacity .3s ease-in;
  transition: opacity .3s ease-in;
  display: none;
}

<div
    className="c0"
  />,
  .c0 {
  background-color: #ffffff;
  color: #000000;
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  width: 40%;
  height: medium;
  -webkit-transform: translateX(100%);
  -ms-transform: translateX(100%);
  transform: translateX(100%);
  -webkit-transition: -webkit-transform 0.2s ease-out;
  -webkit-transition: transform 0.2s ease-out;
  transition: transform 0.2s ease-out;
  overflow-x: hidden;
  overflow-y: scroll;
  box-shadow: 0px 0px 10px rgba(0,0,0,0.12);
}

<div
    className="c0"
    color="primary"
    size="medium"
  />,
]
`;
